src/EVM/Expr.hs

```diff
@@ -1028,30 +1059,39 @@ evalProp prop =
   in if (new == prop) then prop else evalProp new
   where
     go :: Prop -> Prop
+    -- rewrite everything as LEq or LT
+    go (PGEq a b) = PLEq b a
+    go (PGT a b) = PLT b a
+
+    -- LT/LEq comparisions
+    go (PLT  (Var _) (Lit 0)) = PBool False
+    go (PLEq (Lit 0) (Var _)) = PBool True
+    go (PLT  (Lit val) (Var _)) | val == maxLit = PBool False
+    go (PLEq (Var _) (Lit val)) | val == maxLit = PBool True
     go (PLT (Lit l) (Lit r)) = PBool (l < r)
-    go (PGT (Lit l) (Lit r)) = PBool (l > r)
-    go (PGEq (Lit l) (Lit r)) = PBool (l >= r)
     go (PLEq (Lit l) (Lit r)) = PBool (l <= r)
+
+    -- negations
     go (PNeg (PBool b)) = PBool (Prelude.not b)
+    go (PNeg (PNeg a)) = a
 
+    -- And/Or
     go (PAnd (PBool l) (PBool r)) = PBool (l && r)
     go (PAnd (PBool False) _) = PBool False
     go (PAnd _ (PBool False)) = PBool False
-
-    go (POr (PBool l) (PBool r)) = PBool (l || r)
     go (POr (PBool True) _) = PBool True
     go (POr _ (PBool True)) = PBool True
+    go (POr (PBool l) (PBool r)) = PBool (l || r)
 
+    -- Imply
     go (PImpl (PBool l) (PBool r)) = PBool ((Prelude.not l) || r)
```

Comment:
I know it's not part of this PR, but I'm curious what benefit we get from this? Is it easier for the solver? Or does it make constant folding more powerful?

Reply:
Pushed as 92638ffb702592d77d4886be7ac8bf6f1982f222:

```
     -- Imply
-    go (PImpl (PBool l) (PBool r)) = PBool ((Prelude.not l) || r)
+    go (PImpl _ (PBool True)) = PBool True
     go (PImpl (PBool True) b) = b
     go (PImpl (PBool False) _) = PBool True
```

Basically, True is implied by everything. This means that for all combinations of `(PBool a)` and `(PBool b)`, the original rewrite is now useless. So are now more generic! Cool idea :)

Reply:
Resolving.